### PR TITLE
Workspace info

### PR DESCRIPTION
**Pull request by VermaSh**
This addresses the last TODO in Storage Monitoring Jenkins job #12 
The user can now choose which node they want to run the job on.

### CODE REVIEW

File: Jenkins_jobs/WorkspaceInfo.groovy [hunk 1]
```diff
@@ -1,138 +1,100 @@
 @Library('NodeHelper') _
 import jenkins.model.Jenkins;
 import hudson.model.Computer;
+import org.jenkinsci.plugins.workflow.steps.TimeoutStep;
```
Comment (AdamBrousseau) [inline Jenkins_jobs/WorkspaceInfo.groovy]: Why is this needed as an import? `timeout` is a regular pipeline step.

File: Jenkins_jobs/WorkspaceInfo.groovy [hunk 2]
```diff
@@ -1,138 +1,100 @@
...
 
 clones = [:]
-List<String> newMachines = new ArrayList<String>();
 
 /* Iterates over all the online nodes in Jenkins and
  * prints contents of workspace folder along with
  * the space they occupy. With exception of tmp directory
```
Comment (AdamBrousseau) [inline Jenkins_jobs/WorkspaceInfo.groovy]: I don't understand the comment about the tmp dir

File: Jenkins_jobs/WorkspaceInfo.groovy [hunk 3]
```diff
@@ -158,10 +124,47 @@ def beautify(machineName, workspaceDirectory, workspaceStats, subdirectories) {
...
+        workspaceDirectory,
+        workspaceStatscmd,
+        subdirectoryStatscmd,
+        statOutputHeading
+        ) {
+
+    String currentMachine = env.NODE_NAME
+    clones[machineName] = {
+        try {
+            timeout(time: timeOut.getTime(), unit: 'HOURS') {
```
Comment (AdamBrousseau) [inline Jenkins_jobs/WorkspaceInfo.groovy]: I don't think we need all this fancy timeout stuff, I think we can just ref the param here

File: Jenkins_jobs/WorkspaceInfo.groovy [hunk 4]
```diff
@@ -1,138 +1,100 @@
...
+            if (currentInstance != null) {
+                computers = currentInstance.getComputers();
+            }
         }
 
+
         for (Computer computer : computers) {
             String machineName = computer.getName();
 
             if (computer.isOnline() && computer.getName() != "") {
```
Comment (AdamBrousseau) [inline Jenkins_jobs/WorkspaceInfo.groovy]: `computer.getName()` already stored in `machineName`

File: Jenkins_jobs/WorkspaceInfo.groovy [hunk 5]
```diff
@@ -1,138 +1,100 @@
...
 clones = [:]
-List<String> newMachines = new ArrayList<String>();
 
 /* Iterates over all the online nodes in Jenkins and
  * prints contents of workspace folder along with
  * the space they occupy. With exception of tmp directory
+ * 
+ * The nodes can be either passed in or grabbed from Jenkins
  */
 node {
```
Comment (AdamBrousseau) [inline Jenkins_jobs/WorkspaceInfo.groovy]: Does any of this need to be on a `node`?

File: Jenkins_jobs/WorkspaceInfo.groovy [hunk 6]
```diff
@@ -158,10 +124,47 @@ def beautify(machineName, workspaceDirectory, workspaceStats, subdirectories) {
...
+def setupParallelPipelines(
+        timeOut,
+        machineName,
+        workspaceDirectory,
+        workspaceStatscmd,
+        subdirectoryStatscmd,
+        statOutputHeading
+        ) {
+
+    String currentMachine = env.NODE_NAME
```
Comment (AdamBrousseau) [inline Jenkins_jobs/WorkspaceInfo.groovy]: Seems uneeded

File: Jenkins_jobs/WorkspaceInfo.groovy [hunk 7]
```diff
@@ -158,10 +124,47 @@ def beautify(machineName, workspaceDirectory, workspaceStats, subdirectories) {
...
+                            script: subdirectoryStatscmd,
+                            returnStdout: true).trim()
+
+                    println beautify(statOutputHeading, workspaceDirectory, workspaceStats,subdirectories)
+                }
+            }
+        } catch(err) {
+            error (errorMessage(machineName, err.message))
+        } finally {
+            cleanWs()
```
Comment (AdamBrousseau) [inline Jenkins_jobs/WorkspaceInfo.groovy]: Seems we are only catching errors so we can cleanup and print them in a more pretty way? What are we cleaning up?

File: Jenkins_jobs/WorkspaceInfo.groovy (hunk 2 above)
Comment (VermaSh) [reply Jenkins_jobs/WorkspaceInfo.groovy]: tmp directory is any directory with "@tmp" in the name. I'll update the comment

File: Jenkins_jobs/WorkspaceInfo.groovy (hunk 4 above)
Comment (VermaSh) [reply Jenkins_jobs/WorkspaceInfo.groovy]: This is in case `machineName` has the domain but the computer on Jenkins doesn't have the domain in it's name. If I use `machineName` it won't work. Hence, `computer.getName()`

File: Jenkins_jobs/WorkspaceInfo.groovy (hunk 7 above)
Comment (VermaSh) [reply Jenkins_jobs/WorkspaceInfo.groovy]: Originally the idea was to print errors at the end, as a summary. But ran into issues with it so now it's printing the error with a keyword to make it easier to `Crtl+f` the log for errors. There's also the Blue Ocean plug-in if they don't want to sift through the log for errors.

The clean up is anything that may have been put on the disk during the execution. Not sure what but it's there just for good measures.

File: Jenkins_jobs/WorkspaceInfo.groovy (hunk 4 above)
Comment (AdamBrousseau) [reply Jenkins_jobs/WorkspaceInfo.groovy]: I don't understand
```
machineName = computer.getName()
```
So why have `if computer.getName()` when you already have it in a variable?

File: Jenkins_jobs/WorkspaceInfo.groovy [hunk 8]
```diff
@@ -3,136 +3,96 @@ import jenkins.model.Jenkins;
...
-List<String> newMachines = new ArrayList<String>();
 
 /* Iterates over all the online nodes in Jenkins and
  * prints contents of workspace folder along with
  * the space they occupy. With exception of tmp directory
+ * tmp directory is any directory with "@tmp" in the name
+ * 
+ * The nodes can be either passed in or grabbed from Jenkins
  */
 node {
```
Comment (AdamBrousseau) [inline Jenkins_jobs/WorkspaceInfo.groovy]: I don't think you understood my previous question. Do we need a node block to do this?

File: Jenkins_jobs/WorkspaceInfo.groovy [hunk 9]
```diff
@@ -141,6 +101,10 @@ node {
 
         cleanWs()
     }
+
+    stage('Gather_Space_Info') {
+        parallel clones
```
Comment (AdamBrousseau) [inline Jenkins_jobs/WorkspaceInfo.groovy]: Running the "clones" while still occupying the original node is a perfect recipe for deadlock. If we need the node block thats fine, but release the node before running the parallel.

File: Jenkins_jobs/WorkspaceInfo.groovy (hunk 7 above)
Comment (AdamBrousseau) [reply Jenkins_jobs/WorkspaceInfo.groovy]: My point is, if all we're doing is making the error message pretty, and the cleanWs isn't really doing anything, then we don't really need the try-catch-finally at all.

File: Jenkins_jobs/WorkspaceInfo.groovy (hunk 4 above)
Comment (VermaSh) [reply Jenkins_jobs/WorkspaceInfo.groovy]: ah yes, you are right that is redundant. I completely missed the first if condition which validates the computer names. I'll take it out

File: Jenkins_jobs/WorkspaceInfo.groovy (hunk 9 above)
Comment (VermaSh) [reply Jenkins_jobs/WorkspaceInfo.groovy]: ahh yes, that's true. I'll move clones out of the node
